### ISSUE TITLE
got ERROR: (gcloud.auth.docker-helper) on local build

### ISSUE THREAD

**Issue by dfang** [GoogleContainerTools/skaffold#4931]
```
DEBU[0000] Running docker build: context: ., dockerfile: Dockerfile 
ERROR: (gcloud.auth.docker-helper) 
This may be due to network connectivity issues. Please check your network settings, and the status of the service you are trying to reach.
DEBU[0307] Found dependencies for dockerfile: [{go.mod /src true} {go.sum /src true} {main.go /src true}] 
Sending build context to Docker daemon  58.88kB
Step 1/12 : FROM golang:1.12.5 as builder
 ---> 1ef078f0da9e
Step 2/12 : WORKDIR /src
 ---> Using cache
 ---> 348fb407a236
Step 3/12 : COPY go.mod go.sum ./
 ---> Using cache
 ---> b729d9832492
Step 4/12 : ARG SKAFFOLD_GO_GCFLAGS
 ---> Using cache
 ---> e798bea9793e
Step 5/12 : RUN go mod download
 ---> Running in 5ce87a958404
```

### Expected behavior
no this error if code changes when `skaffold dev`

### Actual behavior

It hangs at first line for a couple of minutes, here is my docker config.json

```
$ cat ~/.docker/config.json 
{
  "experimental" : "disabled",
  "auths" : {
    "docker.pkg.github.com" : {

    },
    "https://index.docker.io/v1/" : {

    }
  },
  "HttpHeaders" : {
    "User-Agent" : "Docker-Client/19.03.13 (darwin)"
  },
  "stackOrchestrator" : "swarm",
  "credsStore" : "desktop"
}
```

### Information

- Skaffold version: v1.15.0
- Operating system: macOS 10.15.7
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v2beta8
kind: Config
metadata:
  name: tenant-manager
build:
  local: {}
  artifacts:
  - image: tenant-manager
deploy:
  kubectl:
    manifests:
    - k8s/deployment.yaml
    - k8s/svc.yaml
    - k8s/postgres.yaml
    - k8s/hasura.yaml
```

### Steps to reproduce the behavior

1.  https://github.com/GoogleContainerTools/skaffold/tree/master/examples/getting-started
2. `skaffold dev -vdebug`
3.  both my app and this example hangs 

### FYI

i don't have any [credHelpers section](https://stackoverflow.com/questions/59148598/how-to-get-gcloud-auth-docker-helper-out-of-the-habit-to-hook-into-non-related-d) in my ~/.docker/config.json,  and docker info log as below

```
 λ  docker info
Client:
 Debug Mode: false

Server:
 Containers: 212
  Running: 78
  Paused: 0
  Stopped: 134
 Images: 192
 Server Version: 19.03.13
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 8fba4e9a7d01810a393d5d25a3621dc101981175
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.76-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 1.944GiB
 Name: docker-desktop
 ID: ZBT3:OZFW:HMUQ:6O3M:4WFG:SWLZ:YPXB:QKLM:VA7P:JPC7:454I:DQ7I
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: gateway.docker.internal:3128
 HTTPS Proxy: gateway.docker.internal:3129
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Registry Mirrors:
  https://mirror.baidubce.com/
  https://hub-mirror.c.163.com/
 Live Restore Enabled: false
 Product License: Community Engine
```

**Comment by dfang** [comment]
when i change 

```
build:
  local: {}
```
to

```
build:
  local: 
    useDockerCLI: true
```

it runs fast, but i see a line of log `DEBU[0000] Running command: [/usr/local/bin/minikube profile list -o json] `,  that confused me 

```
 λ  skaffold dev -vdebug
INFO[0000] starting gRPC server on port 50051           
INFO[0000] starting gRPC HTTP server on port 50052      
INFO[0000] Skaffold &{Version:v1.15.0 ConfigVersion:skaffold/v2beta8 GitVersion: GitCommit:c4dce04b0d692dce61051d14cf4a2c04c8c688bd GitTreeState:dirty BuildDate:2020-10-02T16:35:42Z GoVersion:go1.15.2 Compiler:gc Platform:darwin/amd64} 
DEBU[0000] Update check not enabled, skipping.          
INFO[0000] Loaded Skaffold defaults from "/Users/mj/.skaffold/config" 
DEBU[0000] could not parse date ""                      
INFO[0000] Using kubectl context: docker-desktop        
DEBU[0000] Using builder: local                         
DEBU[0000] Running command: [/usr/local/bin/minikube profile list -o json] 
DEBU[0001] Command output: [{"invalid":[],"valid":[{"Name":"minikube","Status":"","Config":{"Name":"minikube","KeepContext":false,"EmbedCerts":false,"MinikubeISO":"","KicBaseImage":"kicbase/stable:v0.0.12-snapshot3@sha256:1d687ba53e19dbe5fafe4cc18aa07f269ecc4b7b622f2251b5bf569ddb474e9b","Memory":1991,"CPUs":2,"DiskSize":20000,"VMDriver":"","Driver":"docker","HyperkitVpnKitSock":"","HyperkitVSockPorts":[],"DockerEnv":null,"ContainerVolumeMounts":null,"InsecureRegistry":null,"RegistryMirror":["https://registry.docker-cn.com"],"HostOnlyCIDR":"192.168.99.1/24","HypervVirtualSwitch":"","HypervUseExternalSwitch":false,"HypervExternalAdapter":"","KVMNetwork":"default","KVMQemuURI":"qemu:///system","KVMGPU":false,"KVMHidden":false,"DockerOpt":null,"DisableDriverMounts":false,"NFSShare":[],"NFSSharesRoot":"/nfsshares","UUID":"","NoVTXCheck":false,"DNSProxy":false,"HostDNSResolver":true,"HostOnlyNicType":"virtio","NatNicType":"virtio","KubernetesConfig":{"KubernetesVersion":"v1.18.3","ClusterName":"minikube","APIServerName":"minikubeCA","APIServerNames":null,"APIServerIPs":null,"DNSDomain":"cluster.local","ContainerRuntime":"docker","CRISocket":"","NetworkPlugin":"","FeatureGates":"","ServiceCIDR":"10.96.0.0/12","ImageRepository":"registry.cn-hangzhou.aliyuncs.com/google_containers","LoadBalancerStartIP":"","LoadBalancerEndIP":"","ExtraOptions":null,"ShouldLoadCachedImages":true,"EnableDefaultCNI":false,"CNI":"","NodeIP":"","NodePort":8443,"NodeName":""},"Nodes":[{"Name":"","IP":"172.17.0.2","Port":8443,"KubernetesVersion":"v1.18.3","ControlPlane":true,"Worker":true},{"Name":"m02","IP":"172.17.0.3","Port":0,"KubernetesVersion":"v1.18.3","ControlPlane":false,"Worker":true},{"Name":"m03","IP":"172.17.0.4","Port":0,"KubernetesVersion":"v1.18.3","ControlPlane":false,"Worker":true}],"Addons":{"ambassador":false,"csi-hostpath-driver":false,"dashboard":true,"default-storageclass":true,"efk":false,"freshpod":false,"gcp-auth":false,"gvisor":false,"helm-tiller":false,"ingress":false,"ingress-dns":false,"istio":false,"istio-provisioner":false,"kubevirt":false,"logviewer":false,"metallb":false,"metrics-server":false,"nvidia-driver-installer":false,"nvidia-gpu-device-plugin":false,"olm":false,"pod-security-policy":false,"registry":false,"registry-aliases":false,"registry-creds":false,"storage-provisioner":true,"storage-provisioner-gluster":false,"volumesnapshots":false},"VerifyComponents":{"apiserver":true,"system_pods":true},"StartHostTimeout":360000000000}}]}] 
DEBU[0001] setting Docker user agent to skaffold-v1.15.0 
DEBU[0001] push value not present, defaulting to false because localCluster is true 
```

**Comment by dfang** [comment]
after some debugging, the problem is here, 
https://github.com/docker/cli/blob/75c1bb1f33d7cedbaf48404597d5bf9818199480/cli/command/image/build.go#L364

if i comment out this line, and L204, `skaffold build` and `skaffold dev` will run without any hesitation

**Comment by dfang** [comment]
possible fix here is:

1.  check if it's local build and push set to false,  if it's true, skip GetAllAuthConfigs

a [builder](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/local/local.go#L40) has information about  if it's local, and push false settings, however it's hard to get builder in [image.go](https://github.com/GoogleContainerTools/skaffold/blob/12d64a2f75437d7d1b904c39e9b1010a74a3a91e/pkg/skaffold/docker/image.go#L184)


2. or update github.com/docker/cli

it seems skaffold imports this version of [docker/cli api](https://github.com/docker/cli/tree/ef2f64abbd37edfa148f745fa0013731b5074d1b),  maybe update it, but  there are credHelpers issues on docker cli too

https://github.com/docker/cli/issues/1989

https://github.com/docker/cli/issues/1591

**Comment by dfang** [comment]
this issue is related to https://github.com/GoogleContainerTools/skaffold/issues/4724

both `useBuildkit: true` and `useDockerCLI: true` in `skaffold.yaml` will skip ` authConfigs, _ := DefaultAuthHelper.GetAllAuthConfigs(ctx) `

since buildkit's goal is to separate out docker build experience into a separate project, allowing different users to collaborate on the underlying technology and reuse and customize it in different ways.  and [BuildKit has been integrated to docker build since Docker 18.09](https://docs.docker.com/develop/develop-images/build_enhancements/), maybe it's time to switched to buildKit by default ?

**Comment by MarlonGamez** [comment]
Hi @dfang , thanks for opening this and looking into it so deeply! 

We currently have a PR in the works that might help with this issue. https://github.com/GoogleContainerTools/skaffold/pull/4597
I wonder if only computing select auth would give much of a speed up for you

**Comment by tstromberg** [comment]
Should this be closed in preference to #4724 ?

**Comment by dfang** [comment]
> Hi @dfang , thanks for opening this and looking into it so deeply!
> 
> We currently have a PR in the works that might help with this issue. #4597
> I wonder if only computing select auth would give much of a speed up for you

yes, here is how i chceked:

```
gh repo clone GoogleContainerTools/skaffold
cd skaffold
gh pr checkout 4597
GOPATH=~/go make install

in project root:
~/go/bin/skaffold dev
```
